packages/astro/src/core/render/core.ts

```diff
@@ -35,6 +35,29 @@ export async function getParamsAndProps(
 			const paramsMatch = route.pattern.exec(pathname);
 			if (paramsMatch) {
 				params = getParams(route.params)(paramsMatch);
+
+				// If we have an endpoint at `src/pages/api/[slug].ts` that's prerendered, and the `slug`
+				// is `undefined`, throw an error as we can't generate the `/api` file and `/api` directory
+				// at the same time. Using something like `[slug].json.ts` instead will work.
+				if (route.type === 'endpoint' && mod.getStaticPaths) {
+					const lastSegment = route.segments[route.segments.length - 1];
+					// Check last segment is solely `[slug]` or `[...slug]` case (dynamic). Make sure it's not
+					// `foo[slug].js` by checking segment length === 1. Also check here if the params are undefined.
+					if (
+						lastSegment.length === 1 &&
+						lastSegment[0].dynamic &&
+						Object.values(params).some((v) => v === undefined)
```

Comment:
TODO: I think this could be improved a little to check the params of the last segment only. Maybe assuming `params` is ordered and getting the last value could be enough.